internal/provision/profiles.go

```diff
@@ -34,81 +37,156 @@ const (
 	}
 
 	files, err := os.ReadDir(absPath)
 	if err != nil {
-		return errors.NewCommonEdgeX(errors.KindServerError, "failed to read directory", err)
+		return nil, errors.NewCommonEdgeX(errors.KindServerError, "failed to read directory for profiles", err)
 	}
 
 	if len(files) == 0 {
-		return nil
+		return nil, nil
 	}
 
-	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
 	lc.Infof("Loading pre-defined profiles from %s(%d files found)", absPath, len(files))
-
-	var addProfilesReq []requests.DeviceProfileRequest
-	dpc := bootstrapContainer.DeviceProfileClientFrom(dic.Get)
+	var addProfilesReq, processedProfilesReq []requests.DeviceProfileRequest
 	for _, file := range files {
-		var profile dtos.DeviceProfile
 		fullPath := filepath.Join(absPath, file.Name())
-		if strings.HasSuffix(fullPath, yamlExt) || strings.HasSuffix(fullPath, ymlExt) {
-			content, err := os.ReadFile(fullPath)
-			if err != nil {
-				lc.Errorf("Failed to read %s: %v", fullPath, err)
-				continue
-			}
+		processedProfilesReq, edgexErr = processProfiles(fullPath, fullPath, nil, lc, dpc)
+		if edgexErr != nil {
+			lc.Error(edgexErr.Error())
+			return addProfilesReq, nil
+		}
+		if len(processedProfilesReq) > 0 {
+			addProfilesReq = append(addProfilesReq, processedProfilesReq...)
+		}
+	}
+	return addProfilesReq, nil
+}
 
-			err = yaml.Unmarshal(content, &profile)
-			if err != nil {
-				lc.Errorf("Failed to YAML decode profile %s: %v", file.Name(), err)
-				continue
-			}
-		} else if strings.HasSuffix(fullPath, jsonExt) {
-			content, err := os.ReadFile(fullPath)
-			if err != nil {
-				lc.Errorf("Failed to read %s: %v", fullPath, err)
-				continue
-			}
+func loadProfilesFromURI(inputURI string, parsedURI *url.URL, dpc interfaces.DeviceProfileClient, secretProvider bootstrapInterfaces.SecretProvider, lc logger.LoggingClient) ([]requests.DeviceProfileRequest, errors.EdgeX) {
+	var edgexErr errors.EdgeX
+	bytes, err := file.Load(inputURI, secretProvider, lc)
+	if err != nil {
+		return nil, errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to load Profile list from URI %s", parsedURI.Redacted()), err)
+	}
 
-			err = json.Unmarshal(content, &profile)
-			if err != nil {
-				lc.Errorf("Failed to JSON decode profile %s: %v", file.Name(), err)
-				continue
-			}
-		} else {
+	if len(bytes) == 0 {
+		return nil, nil
+	}
+
+	var files []string
+	err = json.Unmarshal(bytes, &files)
+	if err != nil {
+		return nil, errors.NewCommonEdgeX(errors.KindServerError, "could not unmarshal Profile list contents", err)
+	}
+	if len(files) == 0 {
+		return nil, nil
+	}
+
+	lc.Infof("Loading pre-defined profiles from %s(%d files found)", parsedURI.Redacted(), len(files))
+	var addProfilesReq, processedProfilesReq []requests.DeviceProfileRequest
+	for _, file := range files {
+		fullPath, redactedPath := GetFullAndRedactedURI(parsedURI, file, "profile", lc)
+		if fullPath == "" || redactedPath == "" {
```

Comment:
Do we need this check? When would it ever be true?

Reply:
no..removing